### ISSUE TITLE
Character encoding problems

### ISSUE THREAD

**Issue by raveren** [kint-php/kint#172]
Hello, author of Kint here :)

If you are reading this, it might mean you came here having issues with wrong display of characters in your (native?) language. I need _your_ help in solving this!

The problem is **extremely** difficult to solve _universally_ and so I humbly ask you to send me the file that displays characters incorrectly. 

You can delete everything else from said file, I basically just need the dumped characters. 

Also, chances are the content that gets dumped comes from an outside source - whether it be a file, a database call or an outside request. If that's the case, I just need that precious garbled string - you can save it via [file_put_contents()](http://php.net/manual/en/function.file-put-contents.php). I'd appreciate to receive the file verbatim, but if it has private information, please `str_replace` it before dumping (but make sure the resulting string still has display issues in Kint.).

If you send such file to me [raveren@gmail.com] and it helps me, I will credit you and you have my gratitude either way :)

**Edit by @jnvsor:** Since I'm the maintainer now, you should probably send them to me

**Comment by lkraav** [comment]
I'm trying to record clean text-only browser request processing backtraces with `error_log( @s( 1 ) )`. This results in bunch of accompanying HTML getting output, like `<pre>` tags etc.

Next I tried `error_log( @~s( 1 ) )` as shown in #139. HTML went away, but a bunch of badly encoded junk was spit on the terminal screen.

Only this works:

```
Kint::enabled( Kint::MODE_WHITESPACE );
error_log( @s( 1 ) );
```

I would really like to have this with modifiers only. Documentation doesn't seem to show any way to do this. Am I missing something? I'm posting in the encoding ticket because it feels like `@~( 1 )` could be right way, but the terminal output encoding choice is messing up.

**Comment by raveren** [comment]
@lkraav the modifiers are meant for on-the-fly development switches, if your flow is fundamentally different (or you are using Kint for logging or in any other context rather that OTF debugging), you can either write a wrapper function or put the initialization (WHITESPACE output as default) in your bootstrapping process.

Also, this is not really related to character set recognition :)

**Comment by userlond** [comment]
For users megrated from v1 branch into v2.

In Kint 1.x I've used the following code to help him detect string encoding:
`Kint::$charEncodings = array('windows-1251', 'utf8');`

The current version is:
`Kint_Object_Blob::$char_encodings = array('windows-1251', 'utf8');`

**Comment by userlond** [comment]
In latest versions (for ex. 2.2) `Kint_Object_Blob::$char_encodings = array('windows-1251', 'utf8');` is not working. As result all non utf-8 strings displayed as `binary string` with empty content.

![kintphp wrong working with windows-1251 with array](https://www.dropbox.com/s/zkhu602uzgvu62r/screenshot%202018-04-20%20001.png?raw=1)

So something was broken with latest releases.

P.S. If you try to out simple string literal, encoding detection is ok, but content is empty.
To reproduce it:

1. Create `php` file with `windows-1251` encoding
2. Call the code `<? d('привет');`, after Kint library init of courder (it should be some cyrillic string, typed into windows-1251 aka cp-1251 encoding)
3. The content  (server: debian 7, php 5.3.3; client windows 7, chrome 63) is like on screenshot below

![kintphp wrong working with windows-1251 with literal](https://www.dropbox.com/s/1y2hi6pmha30xme/screenshot%202018-04-20%20002.png?raw=1)

**Comment by jnvsor** [comment]
1251 support was removed before kint 2.0 release because php does *not* do a good job detecting it automatically. Windows-* charsets basically use every byte available so they match almost any binary data. A guess is as good as it gets. Your example string is also the perfectly valid (But obviously wrong) Windows-1252 string `ïðèâåò`

While it should theoretically work if you put it in the `char_encodings` list I make no promises.

Misdetecting things as binary string is probably because you've thrown away the default `char_encodings` of `['ASCII', 'UTF-8']`. They should both come before anything else so you don't accidentally misdetect strings as win-1251 when they're ascii or utf8.

The root cause of your problem seems to be a bug in `mb_detect_encoding` - despite quotation marks being allowed in 1251 it's the quotation marks that make it fail.

I could fix this in the output, but it seems to me that any windows-1251 text with quotation marks in it would fail anyway at the input.

Can you confirm if adding 1251 to the char_encodings with `Kint_Object_Blob::$char_encodings[] = 'windows-1251';` instead solves at least some of your problems?

**Comment by userlond** [comment]
Thanks fo reply. Unfortunately no issues were solved with `Kint_Object_Blob::$char_encodings[] = 'windows-1251';`. Temporary reverted to version 1.1, now windows-1251 strings are showed ok.

I think this thread needs more comments of people using non-standard encodings to identify issue importance.
![kintphp 1.1 outputs window-1251 strings ok](https://www.dropbox.com/s/b2q99sci2k1d4pz/screenshot%202018-04-24%20001.png?raw=1)

**Comment by jnvsor** [comment]
I could add a flag that searches for "Generic windows-12xx encoding" and only disallows certain control characters (Like I have for 1252, but more flexible) - but it's a hail-mary and there's a *large* chance this will misidentify strings if you have multiple such encodings.

@userlond Do you work with any other win-125x or iso-8859-x encodings besides win-1251 when you're using kint?

**Comment by userlond** [comment]
Projects I've worked on use  `win-1251` and `utf-8`, no other `win-125x` or `iso-8859-x` mixed together.

I think that to decide patching we need more use cases from other `kintphp` users. I thought there is some bug in 2.x , because 1.x branch working ok for me, but if there is some logical difficulties with encodings, patching must be careful to prevent troubles for other users.

P.S. Some googling showed that problem of encoding detection is very complex. What if give user ability to set this function, i.e. make a parameter to provide own callable to detect econding, something like:

```php
<?php
// for php >=5.3
\Kint::$detectEncoding = function($e) {
    // some magic
};

// for php <=5.2
function magic($e) {
    // some magic
};
\Kint::$detectEncoding = 'magic';
```
And by default use existing method.

**Comment by jnvsor** [comment]
If there's literally no way to do it, I'd rather not give a custom callback that pretends there is. If there is I'd rather implement it inside kint.

I might be able to make a *very* iffy way to get those old charsets to work based on the old iconv code, I'll ping you when I have it working

**Comment by jnvsor** [comment]
@userlond can you try af26fda39671603a4342c10997ebd2c1f1a1935a from [my local test branch](https://github.com/jnvsor/kint/commits/test)? The theory is that you can put ambiguous single byte charsets in `BlobObject::$legacy_encodings` and at least rule out the incorrect ones with iconv and a control character preg match.

**Comment by userlond** [comment]
Simple test with `windows-1251` encoding works perfectly on php5.3:
```php
<?php
// this php file has windows-1251 endoding

// include composer somehow

Kint\Object\BlobObject::$legacy_encodings = array('windows-1251');
d('привет');
d(
    array(
        array('id' => 1, 'name' => 'Спутник'),
        array('id' => 2, 'name' => 'Балалайка')
    )
);
```
Output:
![cp1251 string in upcoming kintphp](https://www.dropbox.com/s/fjcspo3d4dtxs82/screenshot%202018-04-26%20001.png?raw=1)

**Comment by jnvsor** [comment]
@userlond Great! It also has some basic distinguishing between 1251 and other legacy byte encodings but most of the time a small string will match them all so order is still important. Look forward to the 3.x beta!

**Comment by jnvsor** [comment]
@userlond @raveren I'm very happy with the current implementation on the next branch, and I don't think there's any better way to do it. It should detect encodings as well as possible when configured properly.

In theory you could use frequency or dictionary analysis to be more accurate, but frequency analysis won't work on small strings (which is most of them in kint dumps), and both are huge performance penalties.

If there are no objections, I think I can close this issue when 3.0 releases

**Comment by jnvsor** [comment]
Fixed in 3.0!
